### DISCUSSION TITLE
Package/namespace directive for components

### DISCUSSION THREAD

**Discussion by mitchgrout** [slint-ui/slint#4871] [Ideas] [answered]
Currently, all C++ generated code will be declared within the global scope, so if you have something like...

```js
export component MainWindow inherits Window { ... }
```

...then the generated C++ header will introduce a global-scope type like...

```cpp
class MainWindow { ... };
```

Personally, I like to keep my code rather organized like...

```cpp
namespace project::core { ... }
namespace project::gui  { ... }
namespace project::cli  { ... }
```

...so having my auto-generated code be outside these namespaces makes it a bit awkward. I could do something like...

```cpp
namespace project::gui { using MainWindow = ::MainWindow; }
```

...to pretend it was part of my scope all along, but as a global variable there's always the risk of ODR violations 

One thought could be to have a `package` directive similar to languages like `proto`, e.g.

```js
package project::gui;
export component MainWindow inherits Window { ... }
```

...could lead to generated code like...

```cpp
namespace project::gui {
  class MainWindow { ... };
}
```

For languages like Python, this could be a little strange to implement, since namespacing is achieved through directory structure. I'm not as familiar with how it works in Rust and Typescript, but there could be some challenges there as well. Thoughts?

**Comment by ogoffart** [comment, marked as answer]
Since Slint 1.5, it is possible to specify NAMESPACE in the cmake command to generate the classes in a namespace:
https://slint.dev/releases/1.5.0/docs/cpp/cmake_reference

Example:
https://github.com/slint-ui/slint/blob/a1c283629e6858e2f9b802fdd9c83032ec279066/examples/todo/cpp/CMakeLists.txt#L13

**Reply by mitchgrout** [reply]
Ah, I see this is also exported as `--cpp-namespace` in the 1.5 version of `slint-compiler`
